web3_proxy/src/balance.rs

```diff
@@ -1,80 +1,166 @@
+    stripe_increase_balance_receipt,
+};
 use migration::sea_orm::prelude::Decimal;
-use migration::sea_orm::{DbBackend, DbConn, FromQueryResult, Statement};
-use serde::{Deserialize, Serialize};
+use migration::sea_orm::DbConn;
+use migration::sea_orm::{ColumnTrait, EntityTrait, QueryFilter, QuerySelect};
+use migration::{Func, SimpleExpr};
+use serde::ser::SerializeStruct;
+use serde::Serialize;
 
-/// Implements the balance getter
-#[derive(Clone, Debug, Default, Serialize, Deserialize, FromQueryResult)]
+/// Implements the balance getter which combines data from several tables
+#[derive(Clone, Debug, Default)]
 pub struct Balance {
-    pub user_id: u64,
-    pub total_spent_paid_credits: Decimal,
+    pub admin_deposits: Decimal,
+    pub chain_deposits: Decimal,
+    pub referal_bonus: Decimal,
+    pub referee_bonus: Decimal,
+    pub stripe_deposits: Decimal,
     pub total_spent: Decimal,
-    pub total_deposits: Decimal,
+    pub total_spent_paid_credits: Decimal,
+    pub user_id: u64,
+}
+
+impl Serialize for Balance {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut state = serializer.serialize_struct("balance", 12)?;
+
+        state.serialize_field("admin_deposits", &self.admin_deposits)?;
+        state.serialize_field("chain_deposits", &self.chain_deposits)?;
+        state.serialize_field("referal_bonus", &self.referal_bonus)?;
+        state.serialize_field("referee_bonus", &self.referee_bonus)?;
+        state.serialize_field("stripe_deposits", &self.stripe_deposits)?;
+        state.serialize_field("total_spent", &self.total_spent)?;
+        state.serialize_field("total_spent_paid_credits", &self.total_spent_paid_credits)?;
+        state.serialize_field("user_id", &self.user_id)?;
+
+        state.serialize_field("active_premium", &self.active_premium())?;
+        state.serialize_field("was_ever_premium", &self.was_ever_premium())?;
+        state.serialize_field("balance", &self.remaining())?;
+        state.serialize_field("total_deposits", &self.total_deposits())?;
+
+        state.end()
+    }
 }
 
 impl Balance {
     pub fn active_premium(&self) -> bool {
-        self.was_ever_premium() && self.total_deposits > self.total_spent_paid_credits
+        self.was_ever_premium() && self.total_deposits() > self.total_spent_paid_credits
     }
 
     pub fn was_ever_premium(&self) -> bool {
-        self.user_id != 0 && self.total_deposits >= Decimal::from(10)
+        self.user_id != 0 && self.total_deposits() >= Decimal::from(10)
     }
 
     pub fn remaining(&self) -> Decimal {
-        self.total_deposits - self.total_spent_paid_credits
+        self.total_deposits() - self.total_spent_paid_credits
+    }
+
+    pub fn total_deposits(&self) -> Decimal {
+        self.admin_deposits
+            + self.chain_deposits
+            + self.referal_bonus
+            + self.referee_bonus
```

Comment:
maybe rename to one-time-bonus or so, otherwise its very easy to mix this one up with referral bonus

Reply:
renamed